File: go/registry.go
package cinchor

import (
	"encoding/json"
	"fmt"
	"math/big"
	"regexp"
	"strings"

	omne "github.com/OmneDAO/omne-sdks/go"
	"github.com/cinchor-hq/cinchor-sdk/go/nonce"
)

var hexRe = regexp.MustCompile(`^(0x)?[0-9a-fA-F]+$`)

// ParseAddressReturn decodes a contract address return: hex (0x-prefixed) → om1z;
// non-hex / "0" / nil → "". A short hex is zero-padded to 32 bytes so a
// wrong-width runtime surfaces as a clearly-wrong address, not a silent one.
func ParseAddressReturn(rv any) string {
	if s, ok := rv.(string); ok && hexRe.MatchString(s) {
		hexs := strings.TrimPrefix(s, "0x")
		if len(hexs) > 64 {
			return s
		}
		if len(hexs) < 64 {
			hexs = strings.Repeat("0", 64-len(hexs)) + hexs
		}
		b, err := omne.FromHex(hexs)
		if err != nil {
			return s
		}
		addr, err := omne.ToOmneAddress(b)
		if err != nil {
			return s
		}
		return addr
	}
	if rv == nil {
		return ""
	}
	if s, ok := rv.(string); ok && s == "0" {
		return ""
	}
	return fmt.Sprintf("%v", rv)
}

// toInt64 coerces a JSON-decoded returnValue (float64 for numbers) to int64.
// (Values here — status, spend, timestamps — fit in float64 without loss.)
func toInt64(rv any) int64 {
	switch v := rv.(type) {
	case float64:
		return int64(v)
	case int64:
		return v
	case int:
		return int64(v)
	default:
		return 0
	}
}

// arg builders: ArgAddress can error, so collect the first error while assembling.
type argFn func() (omne.AbiArgument, error)

func aAddr(s string) argFn { return func() (omne.AbiArgument, error) { return omne.ArgAddress(s) } }
func aI64(v int64) argFn   { return func() (omne.AbiArgument, error) { return omne.ArgI64(v), nil } }

func buildArgs(fns ...argFn) ([]omne.AbiArgument, error) {
	out := make([]omne.AbiArgument, len(fns))
	for i, f := range fns {
		a, err := f()
		if err != nil {
			return nil, err
		}
		out[i] = a
	}
	return out, nil
}

func b2i(b bool) int64 {
	if b {
		return 1
	}
	return 0
}

// CapabilityRegistry is a typed wrapper over the deployed accountability
// contract, built on the Omne base SDK's OmneClient.
type CapabilityRegistry struct {
	client   *omne.OmneClient
	contract string
	prefix   string
	gasLimit uint64
	gasPrice string
	nonces   *nonce.Manager // shared nonce authority; nil → base client NextNonce
}

func NewCapabilityRegistry(cfg Config) *CapabilityRegistry {
	gl := cfg.DefaultGasLimit
	if gl == 0 {
		gl = DefaultGasLimit
	}
	gp := cfg.DefaultGasPrice
	if gp == "" {
		gp = DefaultGasPrice
	}
	return &CapabilityRegistry{
		client:   omne.NewOmneClient(cfg.Network.RPCURL, cfg.Network.ChainID),
		contract: cfg.Contract.Address,
		prefix:   ExportPrefixFor(cfg.Contract),
		gasLimit: gl,
		gasPrice: gp,
		nonces:   cfg.Nonces,
	}
}

func (r *CapabilityRegistry) fn(name string) string { return r.prefix + name }

func (r *CapabilityRegistry) send(signer *omne.WalletAccount, method string, args []omne.AbiArgument, gasLimit uint64) (SubmitReceipt, error) {
	if gasLimit == 0 {
		gasLimit = r.gasLimit
	}
	var tx string
	var err error
	if r.nonces != nil {
		// Route through the shared authority: it hands the closure the nonce to
		// sign at (passing &n takes SendContractCall's explicit-nonce branch,
		// bypassing the base client's own NextNonce). Receipt-waiting stays
		// OUTSIDE Submit so the per-signer lock is held only across sign+submit.
		tx, err = r.nonces.Submit(signer.Address, func(n uint64) (string, error) {
			return r.client.SendContractCall(signer, r.contract, r.fn(method), args, big.NewInt(0), gasLimit, r.gasPrice, &n)
		})
	} else {
		tx, err = r.client.SendContractCall(signer, r.contract, r.fn(method), args, big.NewInt(0), gasLimit, r.gasPrice, nil)
	}
	if err != nil {
		return SubmitReceipt{}, err
	}
	// No WaitForReceipt here: the caller's settle() is the single confirmation
	// (it polls committed state directly). The prior redundant 60s@1s receipt
	// wait was pure read amplification stacked under settle's own poll loop —
	// the settle-storm that saturated the node's read path. Return on
	// mempool-accept; settle establishes committed visibility.
	return SubmitReceipt{TransactionHash: tx, Status: "submitted"}, nil
}

// TxReceipt does a single non-blocking receipt check for the async confirmation
// path. included=false means the tx is not yet observable in a block ("not
// found" is expected while it propagates, not an error). success reflects
// whether the tx executed without a hard failure — NOT the contract return code
// (record_action returning a refusal code 2..6 still yields success=true), which
// is why callers confirm the outcome by committed state delta, not this flag.
func (r *CapabilityRegistry) TxReceipt(txHash string) (included, success bool, blockNumber int64, err error) {
	raw, e := r.client.GetTransactionReceipt(txHash)
	if e != nil {
		if strings.Contains(strings.ToLower(e.Error()), "not found") {
			return false, false, 0, nil // still propagating
		}
		return false, false, 0, e
	}
	if len(raw) == 0 || string(raw) == "null" {
		return false, false, 0, nil
	}
	var rec struct {
		Success     bool  `json:"success"`
		BlockNumber int64 `json:"blockNumber"`
	}
	if e := json.Unmarshal(raw, &rec); e != nil {
		return false, false, 0, fmt.Errorf("decode receipt: %w", e)
	}
	return true, rec.Success, rec.BlockNumber, nil
}

// ── writes (signed; require a funded signer) ─────────────────────────
func (r *CapabilityRegistry) MintPermission(signer *omne.WalletAccount, capabilityID, principal, agent string, maxSpend, validUntil int64, allowlistEnabled bool, currentTime int64, gasLimit uint64) (SubmitReceipt, error) {
	args, err := buildArgs(aAddr(capabilityID), aAddr(principal), aAddr(agent), aI64(maxSpend), aI64(validUntil), aI64(b2i(allowlistEnabled)), aI64(currentTime))
	if err != nil {
		return SubmitReceipt{}, err
	}
	return r.send(signer, "mint_permission", args, gasLimit)
}

func (r *CapabilityRegistry) AddAllowedCounterparty(signer *omne.WalletAccount, capabilityID, counterparty string, currentTime int64, gasLimit uint64) (SubmitReceipt, error) {
	key, err := CounterpartyKey(capabilityID, counterparty)
	if err != nil {
		return SubmitReceipt{}, err
	}
	args, err := buildArgs(aAddr(capabilityID), aAddr(key), aI64(currentTime))
	if err != nil {
		return SubmitReceipt{}, err
	}
	return r.send(signer, "add_allowed_counterparty", args, gasLimit)
}

func (r *CapabilityRegistry) UpdatePolicy(signer *omne.WalletAccount, capabilityID string, newMaxSpend, newValidUntil, currentTime int64, gasLimit uint64) (SubmitReceipt, error) {
	args, err := buildArgs(aAddr(capabilityID), aI64(newMaxSpend), aI64(newValidUntil), aI64(currentTime))
	if err != nil {
		return SubmitReceipt{}, err
	}
	return r.send(signer, "update_policy", args, gasLimit)
}

func (r *CapabilityRegistry) RecordAction(signer *omne.WalletAccount, capabilityID string, amountSpent int64, counterparty string, currentTime int64, actionID string, gasLimit uint64) (SubmitReceipt, error) {
	cp := counterparty
	if cp == "" {
		cp = BurnSentinel
	}
	key, err := CounterpartyKey(capabilityID, cp)
	if err != nil {
		return SubmitReceipt{}, err
	}
	args, err := buildArgs(aAddr(capabilityID), aI64(amountSpent), aAddr(key), aI64(currentTime), aAddr(actionID))
	if err != nil {
		return SubmitReceipt{}, err
	}
	return r.send(signer, "record_action", args, gasLimit)
}

func (r *CapabilityRegistry) RevokePermission(signer *omne.WalletAccount, capabilityID string, currentTime int64, gasLimit uint64) (SubmitReceipt, error) {
	args, err := buildArgs(aAddr(capabilityID), aI64(currentTime))
	if err != nil {
		return SubmitReceipt{}, err
	}
	return r.send(signer, "revoke_permission", args, gasLimit)
}

func (r *CapabilityRegistry) RecordAttestation(signer *omne.WalletAccount, attestationID, capabilityID, contextHash string, verdict Verdict, currentTime int64, gasLimit uint64) (SubmitReceipt, error) {
	args, err := buildArgs(aAddr(attestationID), aAddr(capabilityID), aAddr(contextHash), aI64(int64(verdict)), aI64(currentTime))
	if err != nil {
		return SubmitReceipt{}, err
	}
	return r.send(signer, "record_attestation", args, gasLimit)
}

// ── reads (no signer required) ───────────────────────────────────────
func (r *CapabilityRegistry) query(method, id string) (any, error) {
	arg, err := omne.ArgAddress(id)
	if err != nil {
		return nil, err
	}
	res, err := r.client.QueryContract(r.contract, r.fn(method), []omne.AbiArgument{arg}, "")
	if err != nil {
		return nil, err
	}
	return res["returnValue"], nil
}

func (r *CapabilityRegistry) queryInt(method, id string) (int64, error) {
	v, err := r.query(method, id)
	if err != nil {
		return 0, err
	}
	return toInt64(v), nil
}

func (r *CapabilityRegistry) GetStatus(cap string) (int64, error) { return r.queryInt("get_status", cap) }
func (r *CapabilityRegistry) GetMaxSpend(cap string) (int64, error) {
	return r.queryInt("get_max_spend", cap)
}
func (r *CapabilityRegistry) GetValidUntil(cap string) (int64, error) {
	return r.queryInt("get_valid_until", cap)
}
func (r *CapabilityRegistry) GetTotalSpent(cap string) (int64, error) {
	return r.queryInt("get_total_spent", cap)
}
func (r *CapabilityRegistry) GetActionCount(cap string) (int64, error) {
	return r.queryInt("get_action_count", cap)
}
func (r *CapabilityRegistry) GetCreatedAt(cap string) (int64, error) {
	return r.queryInt("get_created_at", cap)
}
func (r *CapabilityRegistry) GetRevokedAt(cap string) (int64, error) {
	return r.queryInt("get_revoked_at", cap)
}
func (r *CapabilityRegistry) GetPolicyVersion(cap string) (int64, error) {
	return r.queryInt("get_policy_version", cap)
}
func (r *CapabilityRegistry) GetAttestationCount(cap string) (int64, error) {
	return r.queryInt("get_attestation_count", cap)
}

// GetAttestationExists is a single-read existence check (one omne_call), used to
// settle Attest cheaply instead of the multi-read GetAttestation fan-out.
func (r *CapabilityRegistry) GetAttestationExists(attestationID string) (bool, error) {
	v, err := r.queryInt("get_attestation_exists", attestationID)
	if err != nil {
		return false, err
	}
	return v > 0, nil
}

func (r *CapabilityRegistry) GetAllowlistEnabled(cap string) (bool, error) {
	v, err := r.queryInt("get_allowlist_enabled", cap)
	return v == 1, err
}

func (r *CapabilityRegistry) IsCounterpartyAllowed(cap, counterparty string) (bool, error) {
	key, err := CounterpartyKey(cap, counterparty)
	if err != nil {
		return false, err
	}
	v, err := r.queryInt("is_counterparty_allowed", key)
	return v == 1, err
}

func (r *CapabilityRegistry) GetAttestation(attestationID string) (AttestationRecord, error) {
	exists, err := r.queryInt("get_attestation_exists", attestationID)
	if err != nil {
		return AttestationRecord{}, err
	}
	chRaw, err := r.query("get_attestation_context_hash", attestationID)
	if err != nil {
		return AttestationRecord{}, err
	}
	pv, err := r.queryInt("get_attestation_policy_version", attestationID)
	if err != nil {
		return AttestationRecord{}, err
	}
	v, err := r.queryInt("get_attestation_verdict", attestationID)
	if err != nil {
		return AttestationRecord{}, err
	}
	t, err := r.queryInt("get_attestation_time", attestationID)
	if err != nil {
		return AttestationRecord{}, err
	}
	return AttestationRecord{
		Exists:        exists == 1,
		ContextHash:   ParseAddressReturn(chRaw),
		PolicyVersion: pv,
		Verdict:       int(v),
		Time:          t,
	}, nil
}

func (r *CapabilityRegistry) GetCapabilityState(cap string) (CapabilityState, error) {
	status, err := r.GetStatus(cap)
	if err != nil {
		return CapabilityState{}, err
	}
	safeAddr := func(method string) string {
		v, err := r.query(method, cap)
		if err != nil {
			return ""
		}
		return ParseAddressReturn(v)
	}
	get := func(f func(string) (int64, error)) (int64, error) { return f(cap) }
	maxSpend, err := get(r.GetMaxSpend)
	if err != nil {
		return CapabilityState{}, err
	}
	validUntil, err := get(r.GetValidUntil)
	if err != nil {
		return CapabilityState{}, err
	}
	totalSpent, err := get(r.GetTotalSpent)
	if err != nil {
		return CapabilityState{}, err
	}
	actionCount, err := get(r.GetActionCount)
	if err != nil {
		return CapabilityState{}, err
	}
	createdAt, err := get(r.GetCreatedAt)
	if err != nil {
		return CapabilityState{}, err
	}
	revokedAt, err := get(r.GetRevokedAt)
	if err != nil {
		return CapabilityState{}, err
	}
	label := "not_found"
	if status >= 0 && int(status) < len(CapabilityStatusLabels) {
		label = CapabilityStatusLabels[status]
	}
	return CapabilityState{
		CapabilityID: cap,
		Status:       CapabilityStatus(status),
		StatusLabel:  label,
		Principal:    safeAddr("get_principal"),
		Agent:        safeAddr("get_agent"),
		MaxSpend:     maxSpend,
		ValidUntil:   validUntil,
		TotalSpent:   totalSpent,
		ActionCount:  actionCount,
		CreatedAt:    createdAt,
		RevokedAt:    revokedAt,
	}, nil
}

File: py/src/cinchor/registry.py
"""Typed wrapper over the deployed accountability contract (port of
capability-registry.ts), built on the Omne base SDK's OmneClient.

Each method maps to a contract export, encoding arguments per the on-chain ABI
and parsing the returned state. The base SDK owns build→sign→submit→poll
(`send_contract_call`) and read (`query_contract`); this layer adds the typed,
contract-qualified surface.

Numeric arguments are encoded as i64 (matching the TS SDK and the contract's
lowered uint128→i64 parameters). Method names are the contract export selectors
`axiom_contract::<name>::<method>` via `export_prefix_for`.
"""

from __future__ import annotations

import re

from omne_sdk import AbiEncode, OmneClient
from omne_sdk.address import to_omne_address

from .config import BURN_SENTINEL, DEFAULT_GAS_LIMIT, DEFAULT_GAS_PRICE, CinchorConfig, export_prefix_for
from .ids import counterparty_key
from .types import AttestationRecord, CapabilityState, CAPABILITY_STATUS_LABELS, SubmitReceipt

_HEX_RE = re.compile(r"^(0x)?[0-9a-fA-F]+$")


def parse_address_return(rv) -> str:
    """Decode a contract address return: hex (0x-prefixed) → om1z; else passthrough.

    The chain is uniformly 32-byte; a short hex is zero-padded to 32 bytes so a
    wrong-width runtime surfaces as a clearly-wrong address, not a silent one.
    """
    if isinstance(rv, str) and _HEX_RE.match(rv):
        hexs = rv[2:] if rv.startswith("0x") else rv
        if len(hexs) > 64:
            return rv
        hexs = hexs.rjust(64, "0")
        try:
            return to_omne_address(bytes.fromhex(hexs))
        except Exception:
            return rv
    if rv == "0" or rv is None:
        return ""
    return str(rv)


def _to_int(rv) -> int:
    return int(rv) if rv is not None else 0


class CapabilityRegistry:
    def __init__(self, client: OmneClient, config: CinchorConfig):
        self._client = client
        self._contract = config.contract.address
        self._prefix = export_prefix_for(config.contract)
        self._gas_limit = config.default_gas_limit or DEFAULT_GAS_LIMIT
        self._gas_price = config.default_gas_price or DEFAULT_GAS_PRICE

    @classmethod
    def connect(cls, config: CinchorConfig) -> "CapabilityRegistry":
        return cls(OmneClient(config.network.rpc_url, config.network.chain_id), config)

    def _fn(self, name: str) -> str:
        return self._prefix + name

    def _send(self, signer, method: str, args, gas_limit=None) -> SubmitReceipt:
        tx = self._client.send_contract_call(
            signer, self._contract, self._fn(method), args,
            value=0, gas_limit=gas_limit or self._gas_limit, gas_price=self._gas_price,
        )
        receipt = self._client.wait_for_receipt(tx, timeout=60.0)
        block = receipt.get("blockNumber") if isinstance(receipt, dict) else None
        return SubmitReceipt(transaction_hash=tx, status="committed" if receipt else "pending", block_number=block)

    # ── writes (signed; require a funded signer) ─────────────────────
    def mint_permission(self, *, signer, capability_id, principal, agent, max_spend,
                        valid_until, allowlist_enabled=False, current_time, gas_limit=None) -> SubmitReceipt:
        return self._send(signer, "mint_permission", [
            AbiEncode.address(capability_id), AbiEncode.address(principal), AbiEncode.address(agent),
            AbiEncode.i64(max_spend), AbiEncode.i64(valid_until),
            AbiEncode.i64(1 if allowlist_enabled else 0), AbiEncode.i64(current_time),
        ], gas_limit)

    def add_allowed_counterparty(self, *, signer, capability_id, counterparty, current_time, gas_limit=None) -> SubmitReceipt:
        key = counterparty_key(capability_id, counterparty)
        return self._send(signer, "add_allowed_counterparty", [
            AbiEncode.address(capability_id), AbiEncode.address(key), AbiEncode.i64(current_time),
        ], gas_limit)

    def update_policy(self, *, signer, capability_id, new_max_spend, new_valid_until, current_time, gas_limit=None) -> SubmitReceipt:
        return self._send(signer, "update_policy", [
            AbiEncode.address(capability_id), AbiEncode.i64(new_max_spend),
            AbiEncode.i64(new_valid_until), AbiEncode.i64(current_time),
        ], gas_limit)

    def record_action(self, *, signer, capability_id, amount_spent, counterparty=None, current_time, gas_limit=None) -> SubmitReceipt:
        key = counterparty_key(capability_id, counterparty or BURN_SENTINEL)
        return self._send(signer, "record_action", [
            AbiEncode.address(capability_id), AbiEncode.i64(amount_spent),
            AbiEncode.address(key), AbiEncode.i64(current_time),
        ], gas_limit)

    def revoke_permission(self, *, signer, capability_id, current_time, gas_limit=None) -> SubmitReceipt:
        return self._send(signer, "revoke_permission", [
            AbiEncode.address(capability_id), AbiEncode.i64(current_time),
        ], gas_limit)

    def record_attestation(self, *, signer, attestation_id, capability_id, context_hash, verdict, current_time, gas_limit=None) -> SubmitReceipt:
        return self._send(signer, "record_attestation", [
            AbiEncode.address(attestation_id), AbiEncode.address(capability_id), AbiEncode.address(context_hash),
            AbiEncode.i64(int(verdict)), AbiEncode.i64(current_time),
        ], gas_limit)

    # ── reads (no signer required) ───────────────────────────────────
    def _query(self, method: str, id_: str):
        r = self._client.query_contract(self._contract, self._fn(method), [AbiEncode.address(id_)])
        return r.get("returnValue")

    def get_status(self, cap): return _to_int(self._query("get_status", cap))
    def get_max_spend(self, cap): return _to_int(self._query("get_max_spend", cap))
    def get_valid_until(self, cap): return _to_int(self._query("get_valid_until", cap))
    def get_total_spent(self, cap): return _to_int(self._query("get_total_spent", cap))
    def get_action_count(self, cap): return _to_int(self._query("get_action_count", cap))
    def get_created_at(self, cap): return _to_int(self._query("get_created_at", cap))
    def get_revoked_at(self, cap): return _to_int(self._query("get_revoked_at", cap))
    def get_policy_version(self, cap): return _to_int(self._query("get_policy_version", cap))
    def get_attestation_count(self, cap): return _to_int(self._query("get_attestation_count", cap))
    def get_allowlist_enabled(self, cap): return _to_int(self._query("get_allowlist_enabled", cap)) == 1

    def is_counterparty_allowed(self, cap, counterparty) -> bool:
        key = counterparty_key(cap, counterparty)
        return _to_int(self._query("is_counterparty_allowed", key)) == 1

    def get_attestation(self, attestation_id) -> AttestationRecord:
        return AttestationRecord(
            exists=_to_int(self._query("get_attestation_exists", attestation_id)) == 1,
            context_hash=parse_address_return(self._query("get_attestation_context_hash", attestation_id)),
            policy_version=_to_int(self._query("get_attestation_policy_version", attestation_id)),
            verdict=_to_int(self._query("get_attestation_verdict", attestation_id)),
            time=_to_int(self._query("get_attestation_time", attestation_id)),
        )

    def get_capability_state(self, cap) -> CapabilityState:
        status = self.get_status(cap)

        def safe_addr(method):
            try:
                return parse_address_return(self._query(method, cap))
            except Exception:
                return ""

        label = CAPABILITY_STATUS_LABELS[status] if 0 <= status < len(CAPABILITY_STATUS_LABELS) else "not_found"
        return CapabilityState(
            capability_id=cap, status=status, status_label=label,
            principal=safe_addr("get_principal"), agent=safe_addr("get_agent"),
            max_spend=self.get_max_spend(cap), valid_until=self.get_valid_until(cap),
            total_spent=self.get_total_spent(cap), action_count=self.get_action_count(cap),
            created_at=self.get_created_at(cap), revoked_at=self.get_revoked_at(cap),
        )
